### DISCUSSION TITLE
Exception handlers for type Exception behave differently than exception handlers for handlers of more specific types

### DISCUSSION THREAD

**Discussion by mdw00d** [fastapi/fastapi#10107] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# app.py
from fastapi import FastAPI
from fastapi.responses import JSONResponse

app = FastAPI()

@app.exception_handler(Exception)
async def some_exception_handler(request, exc):
    print("In exception handler")
    return JSONResponse(
        status_code=422,
        content={"result": "no go"},
    )


@app.get("/")
def read_root():
    raise ValueError("I don't like this value")
    return {"response": "salutations"}

# Test_app.py
from fastapi.testclient import TestClient

from app import app

client = TestClient(app)


def test_heartbeat():
    response = client.get("/")
    assert response.text == '{"result":"no go"}'
```

### Description

In the above code, I would expect the test to succeed, namely, that the JSON response specified in `some_exception_handler` would be returned.   However, instead, the exception is propagated all the way back to the test code as an exception.

However, if the function is decorated using ValueError instead of the general Exception, it works as expected; a JSON response is returned.
```
@app.exception_handler(ValueError)
async def some_exception_handler(request, exc):
```

On a possibly related note, in class FastAPI, method `build_middleware_stack`, when it builds the set of exception handlers to be passed to the middleware, it specifically treats exceptions registered for the class `Exception` separately.  In particular, exceptions registered for the Exception type are passed to error handler instead of exception handler: 
```
        for key, value in self.exception_handlers.items():
            if key in (500, Exception):
                error_handler = value
            else:
                exception_handlers[key] = value
```
this caused an exception handler to not get passed on and invoked as expected (which was the original problem I was trying to solve in my real code before noticing the issue described here).

### Operating System

Windows

### FastAPI Version

0.101.1

### Pydantic Version

2.2.0

### Python Version

3.11.2

**Comment by YuriiMotov** [comment]
Explained in this comment: https://github.com/fastapi/fastapi/discussions/11348#discussioncomment-8912085
